### ISSUE TITLE
Overall error in cross-validation and horizon in performance matrix

### ISSUE THREAD

**Issue by maralbertomar** [facebook/prophet#1347]
Hi everyone!
i'm trying to predict one year data of a weekly time series with cross validation as follow:
`df_raw_cv=cross_validation(m, horizon = '365 days', initial='710 days')`. I obtain this performance matrix:

![image](https://user-images.githubusercontent.com/54624827/75028302-507b9b80-54a0-11ea-87c8-13887126174a.png)

I have two questions about it:
1) I expect that horizon goes on seven by seven days (due to my weekly data), but in certain rows delta is larger (for example, between first and second row there are 28 days), how is it possible?

2) I need to obtain an overall MAPE of the prediction in order to compare the performance with other models. I tried to do the arithmetic average of the MAPEs but due to this variability in delta between horizons, i think this is not properly correct. How do you suggest to calculate it? Is there any built-in function which allow me to calculate an overall error?

Thank you very much

Alberto

**Comment by bletham** [comment]
If you look at the output of `cross_validation` (call it `df_cv`), then you will see that it has a column `ds` that is the timestamp of the point being predicted, and `cutoff` that is the cutoff with which that prediction was made. If you look at `df_cv['ds'] - df_cv['cutoff']`, these are all of the horizon values used for computing performance metrics.

When computing the performance metrics (like MAPE here in particular), it computes the APE for every row in `df_cv`, and then computes MAPE by averaging over a rolling window based on the horizon (`df_cv['ds'] - df_cv['cutoff']`). This is described in the documentation here: https://facebook.github.io/prophet/docs/diagnostics.html and in the code here: https://github.com/facebook/prophet/blob/46e56119835f851714d22b285d2e4081853b9fb1/python/fbprophet/diagnostics.py#L274

I describe the rolling average approach in detail here: https://github.com/facebook/prophet/issues/839#issuecomment-462524968 (what I describe as "what we really want to do" is what we are now doing).

If for each unique value of `df_cv['ds'] - df_cv['cutoff']` you have the same number of rows in df_cv, then I would expect to have equal spacing of horizons in the performance metrics results. I'm guessing that is not the case (that there are some weeks with missing data) and as a result the rolling window sometimes requires averaging over a larger window to reach the minimum data setting. Is that the case?

**Comment by maralbertomar** [comment]
Hi Bletham! Yes, it is correct. I checked the `df_cv` and there were some missing values in those cases. Actually, the real problem in my case is that with weekly data all the horizons calculated  in cross_validation depends to cutoffs. Here a screenshot of my new custom `df_cv` sorted by `delta_horizon = df_cv['ds'] - df_cv['cutoff']` and obtained from `cross_validation(m, horizon = '366 days', initial='366 days'):` 
 
![image](https://user-images.githubusercontent.com/54624827/75243667-9a2ef380-57ca-11ea-95ee-60bb8df59ffa.png)

The first 3 rows, with horizon 2 days, 5 days, 7 days belong to the "first week after cutoff"; they are considered separately only because the cutoffs are in different days of the week but they should be considered as 1 row! 

The possible solution could be to modify the `df['horizon']` in order to calculate the difference between ds and cutoff in my particoular data frequency (week, month, year, etc) or alternatively, to find a way to put your own personalized cutoff dates in cross-validation. What do you think about it?

**Comment by bletham** [comment]
Ah, I see; the issue here is that the command used was
```
cross_validation(m, horizon = '365 days', initial='710 days')
```
and if not specified, the period (the spacing between cutoffs) will default to horizon / 2: https://github.com/facebook/prophet/blob/46e56119835f851714d22b285d2e4081853b9fb1/python/fbprophet/diagnostics.py#L75-L76

which here is 182.5 days. So it will fall on a different day each week. If you specify a `period` that is divisible by 7 (like maybe 140 days, for 20 weeks) then the cutoffs should all fall on the same day of the week.

**Comment by maralbertomar** [comment]
This is true! Thank you Bletham, I solved the problem of my weekly time series but I have still problem with monthly time series. Months don't have the same length, so we cannot find a ` period` standard for every month, do you have any tips about this?

**Comment by bletham** [comment]
There isn't a really good way to do that with months right now because of their variable length. The plan for right now is to enable setting custom cutoffs, which you could then specify at a month-start or month-end frequency as needed by the particular monthly time series. #1351 is the issue for that.

In the meantime, I think the best you could do would be to be to copy out the code for the cross validation and replace them manually right here in a local version of the function:
https://github.com/facebook/prophet/blob/46e56119835f851714d22b285d2e4081853b9fb1/python/fbprophet/diagnostics.py#L108

**Comment by MarcoFusi** [comment]
@bletham we're trying to fix this with pull request #1402
